rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp

```diff
@@ -632,6 +632,144 @@ NodeParameters::declare_parameter(
+    }
+
+    auto name = normalized_namespace + parameter.get_name();
+    // Error if this parameter has already been declared and is different
+    if (__lockless_has_parameter(parameters_, name)) {
+      __undeclare_parameters(parameters_, names);
+      throw rclcpp::exceptions::ParameterAlreadyDeclaredException(
+              "parameter '" + name + "' has already been declared");
+    }
+
+    auto type = parameter.get_type();
+    auto default_value = parameter.get_parameter_value();
+    if (!parameter_descriptor.dynamic_typing) {
+      if (rclcpp::PARAMETER_NOT_SET == type) {
+        type = default_value.get_type();
+      }
+      if (rclcpp::PARAMETER_NOT_SET == type) {
+        __undeclare_parameters(parameters_, names);
+        throw rclcpp::exceptions::InvalidParameterTypeException{
+                name,
+                "cannot declare a statically typed parameter with an uninitialized value"
+        };
+      }
+      parameter_descriptor.type = static_cast<uint8_t>(type);
+    }
+
+    auto result = __declare_parameter_common(
+      name,
+      default_value,
+      parameter_descriptor,
+      parameters_,
+      parameter_overrides_,
+      on_set_parameters_callback_container_,
+      post_set_parameters_callback_container_,
+      &parameter_event,
+      ignore_override);
+
+    // If it failed to be set, then throw an exception.
+    if (!result.successful) {
+      constexpr const char type_error_msg_start[] = "Wrong parameter type";
+      if (
+        0u == std::strncmp(
+          result.reason.c_str(), type_error_msg_start, sizeof(type_error_msg_start) - 1))
+      {
+        __undeclare_parameters(parameters_, names);
+        throw rclcpp::exceptions::InvalidParameterTypeException(name, result.reason);
+      }
+      __undeclare_parameters(parameters_, names);
+      throw rclcpp::exceptions::InvalidParameterValueException(
+              "parameter '" + name + "' could not be set: " + result.reason);
+    }
+
+    names.push_back(name);
+    results.push_back(parameters_.at(name).value);
+  }
+
+  // Publish if events_publisher_ is not nullptr, which may be if disabled in the constructor.
+  if (nullptr != events_publisher_) {
+    parameter_event.node = combined_name_;
+    parameter_event.stamp = node_clock_->get_clock()->now();
+    events_publisher_->publish(parameter_event);
+  }
+
+  return results;
+}
+
+std::vector<rclcpp::ParameterValue>
+NodeParameters::declare_parameters(
+  const std::string & namespace_,
+  const std::vector<rclcpp::Parameter> & parameters,
+  bool ignore_override)
+{
+  std::vector<std::pair<rclcpp::Parameter, rcl_interfaces::msg::ParameterDescriptor>> parameters_;
+  for (const auto & param : parameters) {
+    parameters_.push_back(std::make_pair(param, rcl_interfaces::msg::ParameterDescriptor()));
+  }
+
+  return declare_parameters_atomically(namespace_, parameters_, ignore_override);
```

Comment:
so we are changing all APIs to atomic operation?

Reply:
Only declare_parameters staff.